rfcs/text/typescript.md

```diff
@@ -0,0 +1,259 @@
+    "noImplicitThis": true,
+    "noImplicitAny": false,
+    "noUnusedLocals": true,
+    "outDir": "./lib/",
+    "resolveJsonModule": true,
+    "rootDir": "./src",
+    "skipLibCheck": true,
+    "sourceMap": true,
+    "strict": true,
+    "strictNullChecks": true,
+    "suppressImplicitAnyIndexErrors": true,
+    "target": "es5"
+  },
+  "include": ["src/**/*"]
+}
+```
+
+### Babel config
+
+We will use Babel for transpiling the code and `tsc` for type checking and generating `.d.ts` files. This hybrid approach is common for codebases ported from JavaScript to TypeScript with existing build structures using Babel.
+
+Using Babel enables us to keep listing the exact environments we need to support, i.e. the last 2 versions of our supported browsers. It's also faster to compile than the TypeScript compiler.
+
+We should be able to leave our Babel config as it is as long as type checking is carried out before running Babel in our build setup.
+
+### Packages
+
+The version of the `typescript` package in Carbon should be aligned with the version used in the other Sage TypeScript repositories.
+
+No new packages should need to be added to Carbon for this conversion.
+
+### Scripts
+
+Scripts to generate declaration (`.d.ts`) files and carry out type checking should be added:
+
+```js
+ "scripts": {
+    "type-check": "tsc --noEmit", // Type check the code once
+    "type-check:watch": "npm run type-check -- --watch", // Keep type checking running
+    "build:types": "tsc --emitDeclarationOnly", // Generate declaration files
+  },
+```
+
+The `--extensions '.js','.ts','.tsx'` option should be added to the `babel` script. This will tell Babel to look for TypeScript files and transpile them to JavaScript, as well as transpile any JavaScript files we need.
+
+The `copy-files` script can now ignore TypeScript files as they will be transpiled and copied into the `lib` folder by Babel. `.d.ts` files will be added to the `lib` folder by the `build:types` script, as long as `"outDir": "./lib/",` is specified in the TypeScript config.
+
+The `precompile` script should then be modified to run type checking and generate declaration files, as well as running Babel:
+
+```json
+    "precompile": "npm run type-check && npm run clean-lib && npm run copy-files && npm run build:types && npm run babel",
+```
+
+### Eslint config
+
+Converting to TypeScript means the parser in the eslint config should change to `@typescript-eslint/parser` rather than `babel-eslint`.
+
+A [Pull Request](https://github.com/Sage/carbon/pull/3978) is already in progress to add eslint config for TypeScript in Carbon. Once we convert the library to TypeScript we can simply replace the base config with the overrides specified in that config.
+
+## Components
+
+The the [Basic Example](#basic-example) section for how components will change when converted to TypeScript. The types of each prop will remain the same but defined only in TypeScript rather than both js and ts.
```

Comment:
```suggestion
The [Basic Example](#basic-example) section for how components will change when converted to TypeScript. The types of each prop will remain the same but defined only in TypeScript rather than both js and ts.
```